### PR TITLE
[Limit orders] Execution price implementation

### PR DESCRIPTION
**Pull request by nenadV91**
# Summary

- adds execution price fetching in the background
- when the user clicks on the `market price` button the execution rate will be set as active rate
- if there is no execution price or the execution price is being fetched for new token pair the `market price` button is disabled
Screenshot from 2022-11-04 15-35-12

### To test
- we should basically test all rate related functionality

### CODE REVIEW

**Review by anxolin — COMMENTED** [review]
@nenadV91 why did you decided to implement `useFetchMarketPrice` as a hook instead of as an Updater that we include in the View we want? 

we usually do it like that in other places, right?

File: src/cow-react/modules/limitOrders/updaters/QuoteUpdater/hooks/useQuoteRequestParams.ts [hunk 1]
```diff
@@ -2,30 +2,50 @@ import { useLimitOrdersTradeState } from '@cow/modules/limitOrders/hooks/useLimi
...
+import { getAddress } from '@cow/modules/limitOrders/utils/getAddress'
+import { calculateValidTo } from '@cow/utils/time'
+import { useUserTransactionTTL } from '@src/state/user/hooks'
 
 export function useQuoteRequestParams(): FeeQuoteParams | null {
-  const { inputCurrency, outputCurrency, inputCurrencyAmount, recipient, deadline } = useLimitOrdersTradeState()
+  const { inputCurrency, outputCurrency, recipient, orderKind } = useLimitOrdersTradeState()
   const { chainId, account } = useWeb3React()
+  const { exactTypedValue } = useTypedValue()
+  const [deadline] = useUserTransactionTTL()
```
Comment (shoom3301) [inline src/cow-react/modules/limitOrders/updaters/QuoteUpdater/hooks/useQuoteRequestParams.ts:16]: `useUserTransactionTTL()` - that hook provides data only for Swap. For limit-orders we should use `useLimitOrdersTradeState()`

File: src/cow-react/modules/limitOrders/updaters/QuoteUpdater/hooks/useFetchMarketPrice.ts [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+import { useLimitOrdersTradeState } from '@cow/modules/limitOrders/hooks/useLimitOrdersTradeState'
+import { updateLimitRateAtom } from '@cow/modules/limitOrders/state/limitRateAtom'
+import { useQuoteRequestParams } from './useQuoteRequestParams'
+import { useHandleResponse } from './useHandleResponse'
+import { useHandleError } from './useHandleError'
+
+// Every 10s
+const REFETCH_CHECK_INTERVAL = 10000
+
+export function useFetchMarketPrice() {
```
Comment (shoom3301) [inline src/cow-react/modules/limitOrders/updaters/QuoteUpdater/hooks/useFetchMarketPrice.ts:15]: I think it makes sense to move the hook in the directory:
`hooks/useFetchMarketPrice/index.ts`
And also move `useHandleError` and `useHandleResponse` to the directory

**Review by shoom3301 — APPROVED** [review]

**Review by shoom3301 — APPROVED** [review]

File: src/cow-react/modules/limitOrders/containers/RateInput/index.tsx [hunk 3]
```diff
@@ -1,104 +1,75 @@
...
-    updateLimitRateState({ activeRate: marketRate })
-  }, [marketRate, updateLimitRateState])
+    updateLimitRateState({ activeRate: executionRate, isTypedValue: false })
+  }, [executionRate, updateLimitRateState])
 
   // Handle rate input
   const handleUserInput = useCallback(
     (typedValue: string) => {
-      updateLimitRateState({ activeRate: typedValue })
+      updateLimitRateState({ typedValue, activeRate: toFraction(typedValue, isInversed), isTypedValue: true })
```
Comment (anxolin) [inline src/cow-react/modules/limitOrders/containers/RateInput/index.tsx:44]: I love to see that now, when we parse the amount, we take into account if the user has the `isInverted` flag changed to invert what they type. 

If I understand it correctly, you are now storing the rate in the same units, and we don't invert it when we flip `isInversed`, but we just take it into account when we represent it.

File: src/cow-react/modules/limitOrders/utils/toFirstMeaningfulDecimal.ts [hunk 4]
```diff
@@ -0,0 +1,21 @@
+export function toFirstMeaningfulDecimal(value: string | undefined | null, limit = 6): string {
```
Comment (anxolin) [inline src/cow-react/modules/limitOrders/utils/toFirstMeaningfulDecimal.ts:1]: Why do we need this? 
I believe this is already provided by BigDecimal and so on, and actually this implementations will fail for other locales

File: src/cow-react/modules/limitOrders/utils/toFraction.ts [hunk 5]
```diff
@@ -0,0 +1,16 @@
+import { Fraction } from '@uniswap/sdk-core'
+
+export function toFraction(value: string, isInversed = false) {
+  const [quotient, remainder = ''] = value.split('.')
```
Comment (anxolin) [inline src/cow-react/modules/limitOrders/utils/toFraction.ts:4]: this is not locale friendly. Probably is safer for you to use a library of some sort...

File: src/cow-react/modules/limitOrders/utils/toFraction.ts [hunk 6]
```diff
@@ -0,0 +1,16 @@
...
+
+  if (!value || !Number(value)) return new Fraction(0)
+
+  const numerator = `${quotient}${remainder}`
```
Comment (anxolin) [inline src/cow-react/modules/limitOrders/utils/toFraction.ts:8]: I'm sure we don't need to implement this ourselves and theres libraries we can rely that will do a more solid implementation than this. 

Looks very fragile. Also, what if there's thousands separators here?

File: src/cow-react/modules/limitOrders/updaters/InitialPriceUpdater/index.tsx [hunk 7]
```diff
@@ -0,0 +1,16 @@
+import { useEffect } from 'react'
+import { useUpdateAtom } from 'jotai/utils'
+import { updateLimitRateAtom } from '@cow/modules/limitOrders/state/limitRateAtom'
+import { useGetInitialPrice } from '@cow/modules/limitOrders/hooks/useGetInitialPrice'
+
+// Fetch and update initial price for the selected token pair
+export function InitialPriceUpdater() {
+  const updateLimitRateState = useUpdateAtom(updateLimitRateAtom)
+  const { price, isLoading } = useGetInitialPrice()
```
Comment (anxolin) [inline src/cow-react/modules/limitOrders/updaters/InitialPriceUpdater/index.tsx:9]: Are you sure this is only executed once for the initial price? `useGetInitialPrice` uses a lot of hooks, and would probably be called a few times, right?

**Review by anxolin — COMMENTED** [review]
Great changes, nice to see you found a way. 

I understand it was already hard to figure out this PR, this is why is fine. But I feel this PR could be broken down into different PRs. No need to do it now. just keep in mind to make it smaller and easily reviewable

File: src/cow-react/modules/limitOrders/utils/toFirstMeaningfulDecimal.ts (hunk 4 above)
Comment (nenadV91) [reply src/cow-react/modules/limitOrders/utils/toFirstMeaningfulDecimal.ts:1]: There is `toSignificant` method on Fractions, so we can use that instead

File: src/cow-react/modules/limitOrders/updaters/InitialPriceUpdater/index.tsx (hunk 7 above)
Comment (nenadV91) [reply src/cow-react/modules/limitOrders/updaters/InitialPriceUpdater/index.tsx:9]: The actual `native_price` request happens only once for each currency and then when some currency is changed there is another request for that new currency. And the updater's `useEffect` is called every time a new price or isLoading is changed. But I think I see your point, so I will update the name to `MarketPriceUpdater`

File: src/cow-react/modules/limitOrders/utils/toFraction.ts (hunk 5 above)
Comment (nenadV91) [reply src/cow-react/modules/limitOrders/utils/toFraction.ts:4]: Ok no problem, I've picked this one https://github.com/infusion/Fraction.js but its basically just to parse the input value and then rest is with Fraction from Uniswap

File: src/cow-react/modules/limitOrders/containers/RateInput/index.tsx (hunk 3 above)
Comment (nenadV91) [reply src/cow-react/modules/limitOrders/containers/RateInput/index.tsx:44]: We do take it into account when its represented, but we also have to take it into account when user it typing value and `isInversed` is true.

File: src/cow-react/modules/limitOrders/utils/toFraction.ts (hunk 5 above)
Comment (shoom3301) [reply src/cow-react/modules/limitOrders/utils/toFraction.ts]: You can just use: `split(/\,|\./)`
I don't know any locales that use something besides , and .

**Comment by elena-zh** [comment]
Post-merge review issue #1419

**Pull request merged**